tee_supplicant/src/tee_supplicant.c

```diff
@@ -0,0 +1,296 @@
+	struct suppl_shm *tshm = NULL;
+	size_t offs = 0;
+	size_t sz = 0;
+
+	if (idx >= num_params) {
+		return -1;
+	}
+
+	switch (params[idx].attr & TEE_PARAM_ATTR_TYPE_MASK) {
+	case TEE_PARAM_ATTR_TYPE_MEMREF_INPUT:
+	case TEE_PARAM_ATTR_TYPE_MEMREF_OUTPUT:
+	case TEE_PARAM_ATTR_TYPE_MEMREF_INOUT:
+		break;
+	default:
+		return -2;
+	}
+
+	memset(shm, 0, sizeof(*shm));
+
+	tshm = find_tshm(MEMREF_SHM_ID(params + idx));
+	if (!tshm) {
+		/*
+		 * It doesn't make sense to query required size of an
+		 * input buffer.
+		 */
+		if ((params[idx].attr & TEE_PARAM_ATTR_TYPE_MASK) ==
+		    TEE_PARAM_ATTR_TYPE_MEMREF_INPUT) {
+			return -3;
+		}
+
+		/*
+		 * Buffer isn't found, the caller is querying required size
+		 * of the buffer.
+		 */
+		return 0;
+	}
+
+	sz = MEMREF_SIZE(params + idx);
+	offs = MEMREF_SHM_OFFS(params + idx);
+	if ((sz + offs) < sz) {
+		return -4;
+	}
+	if ((sz + offs) > tshm->size) {
+		return -5;
+	}
+
+	shm->flags = TEEC_MEM_INPUT | TEEC_MEM_OUTPUT;
+	shm->size = sz;
+	shm->id = MEMREF_SHM_ID(params + idx);
+	shm->buffer = (uint8_t *)tshm->p + offs;
+
+	return 0;
+}
+
+static void uuid_from_octets(TEEC_UUID *d, const uint8_t s[TEE_UUID_LEN])
+{
+	d->timeLow = (s[0] << 24) | (s[1] << 16) | (s[2] << 8) | s[3];
+	d->timeMid = (s[4] << 8) | s[5];
+	d->timeHiAndVersion = (s[6] << 8) | s[7];
+	memcpy(d->clockSeqAndNode, s + 8, sizeof(d->clockSeqAndNode));
+}
+
+static int load_ta(uint32_t num_params, struct tee_param *params)
+{
+	int ta_found = 0;
+	size_t size = 0;
+	struct param_value *val_cmd = NULL;
+	TEEC_UUID uuid = { 0 };
+	TEEC_SharedMemory shm_ta = { 0 };
+
+	if (num_params != 2 || get_value(num_params, params, 0, &val_cmd) ||
+	    get_param(num_params, params, 1, &shm_ta)) {
+		return TEEC_ERROR_BAD_PARAMETERS;
+	}
+
+	uuid_from_octets(&uuid, (void *)val_cmd);
+
+	size = shm_ta.size;
+	ta_found = TEECI_LoadSecureModule(&uuid, shm_ta.buffer, &size);
+	if (ta_found != TA_BINARY_FOUND) {
+		LOG_ERR("TA not found");
+		return TEEC_ERROR_ITEM_NOT_FOUND;
+	}
+
+	MEMREF_SIZE(params + 1) = size;
+
+	/*
+	 * If a buffer wasn't provided, just tell which size it should be.
+	 * If it was provided but isn't big enough, report an error.
+	 */
+	if (shm_ta.buffer && size > shm_ta.size) {
+		return TEEC_ERROR_SHORT_BUFFER;
+	}
+
+	return TEEC_SUCCESS;
+}
+
+static int process_request(const struct device *dev)
+{
+	int rc;
+	union tee_supp_msg ts_msg = {
+		.req.num_param = TEE_REQ_PARAM_MAX,
+	};
+
+	rc = receive_request(dev, &ts_msg.req);
+	if (rc) {
+		return rc;
+	}
+
+	LOG_DBG("Receive OPTEE request cmd #%d", ts_msg.req.cmd);
+	switch (ts_msg.req.cmd) {
+	case OPTEE_MSG_RPC_CMD_LOAD_TA:
+		rc = load_ta(ts_msg.req.num_param, ts_msg.req.params);
+		break;
+	default:
+		return TEEC_ERROR_NOT_SUPPORTED;
+	}
+
+	ts_msg.rsp.ret = rc;
+	return send_response(dev, &ts_msg.rsp);
+}
+
+static void tee_supp_main(void *p1, void *p2, void *p3)
+{
+	ARG_UNUSED(p2);
+	ARG_UNUSED(p3);
+	const struct device *dev = p1;
+	int rc = 0;
+
+	while (1) {
```

Comment:
How to stop supplicant?

Reply:
Do we need this?

Reply:
Of course we do. How we can safely stop supplicant before reboot, for example?

Reply:
@lorc, should this comment be fixed? Actually, Zephyr doesn't support shutdown, but it is not a big deal to implement stop function and add atomic variable, as it is made for xenstore/xenconsole if such case is really needed.

Reply:
@firscity Unfortunately this can't be solved by atomic. Thread is blocked suppl_recv call waiting on semaphore similar to the Linux Kernel implementation which is waiting for external signal. I see the 2 valid options to make stop procedure by now.
Hardly stop thread by calling Cancel or to make OP-TEE unblock recv by initiating stop procedure with close_sessions calls. This approach requires additional research because if there are not active session - recv whouldn't be unblocked.
As for me - This requires additional research, because Linux Kernel approach is definitely doesn't fit Zephyr. I suggest leaving todo about this problem due to the lack of time.